### PR TITLE
(WIP) maybeInput and testing

### PR DESCRIPTION
**Pull request by tonyday567**
Trying to apply #4 suggestions:

You can't wrap table tags <td> and <tr> with a div, so there's some refactoring that needs doing if you wanted this type of thing in the library.

I couldn't work out how to write what I had in mind with ApplicativeDo but I got there with normal applicative style.  I'm just at a loss about how to refactor between the two.

I jumped down to a value-level solution just to simplify my ApplicativeDo confusions.

Whilst hacking away, It felt that the Markup bit of an Input could usefully be altered to:

```
data MarkupInput = MarkupInput { inputStyle :: Markup, inputScript :: Markup, inputHtml :: Markup }
```

or even

```
data Widget a = Widget { style :: a, script :: a, html :: a }
```

and then you could inject the css, js and html elements of an input in a neater spot then in the middle of the body.

Similarly, passing around all these Strings that are css styles, html tag wrapping, js code etc, and there should be easy benefits to start using blaze types (or whatever), especially when trying to modify a html element inside an input.  Maybe Text.Blaze is already usable and I didn't notice.

### CODE REVIEW

**Comment by isovector** [comment]
I'm open to PRs if you want to make simplifications in doing things. Most of this code was the lowest effort possible to show it working, but cleaning it up would be awesome!

**Comment by isovector** [comment]
whoops!! replied to this via email and didn't realize it _was_ a PR. sorry! i'll take a look asap

File: src/Web/Suavemente/Input.hs [hunk 1]
```diff
@@ -105,6 +106,15 @@ checkbox label = mkInput parseJSON $ \name v ->
         <input id="{name}" onchange="onChangeFunc(event)" type="checkbox" {bool ("" :: String) "checked='checked'" v} autocomplete="off">
         </td></tr>|]
 
+------------------------------------------------------------------------------
+-- | Create an input driven by an HTML checkbox without table tags.
+checkbox_ :: String -> Bool -> Suave Bool
```
Comment (isovector) [inline src/Web/Suavemente/Input.hs:112]: what do this and `textbox_` buy you?

File: src/Web/Suavemente/Input.hs [hunk 2]
```diff
@@ -152,3 +174,59 @@ enumDropdown
...
+-- | show/hide style string
+display :: Bool -> String
+display b = "display:" ++ bool "none" "block" b
+
+-- | apply a checkbox that turns a Suave a into a Suave (Maybe a)
+maybeInput :: String -> Bool -> Suave a -> Suave (Maybe a)
+maybeInput label start sa =
+  (\c a -> bool Nothing (Just a) c) <$>
+  checkboxShow label "wrap" start <*>
+  div' "wrap" (display start) sa
```
Comment (isovector) [inline src/Web/Suavemente/Input.hs]: you could just add this class to a table row, no?

**Comment by isovector** [comment]
Great start! ~I wonder if there's not a higher-level solution hiding in here somewhere. Explicitly tracking the  visibility DAG seems fraught with peril. Maybe we could just write business as usual Haskell and use something like [data-reify](https://hackage.haskell.org/package/data-reify) to figure out the dependencies.~

I just reread what's going on here and I understand what's happening now. `maybeInput` maybe should be `Suave a -> (a -> Bool) -> Suave b -> Suave (a, Maybe b)` and we can find a solution to serialize the predicate (likely `data-reify` above).

File: src/Web/Suavemente/Input.hs (hunk 1 above)
Comment (tonyday567) [reply src/Web/Suavemente/Input.hs:112]: Just the use of div as a collection or wrapper of inputs.  It looks like <div> can't be used to wrap a <tr>.  See https://stackoverflow.com/questions/23440362/can-we-add-div-inside-table-above-every-tr/23440419. You have to use tbody instead, it seems, but this felt like lock-in to inputs always being one big table.

If the tr and td tags were removed from the input html, you would probably need to keep labels and inputs separate, so formatting like <tr><td>label</td><td>input</td></tr> could eventually be injected. `tableize` for the existing method.

File: src/Web/Suavemente/Input.hs (hunk 2 above)
Comment (tonyday567) [reply src/Web/Suavemente/Input.hs]: I dont think so, no.

**Comment by tonyday567** [comment]
I don't understand, might take me a bit to catch up.

As it stands, (using your above signature), the b always exists, and needs to exist to initialize the input on the js side.  It just hides the b on the haskell side.  So outputting a Suave (a, b) is just as valid.

**Comment by tonyday567** [comment]
In looking at data-reify, I noticed https://github.com/ku-fpg/javascript-bridge by the same group.  It's a useful comparator.

File: src/Web/Suavemente/Input.hs (hunk 2 above)
Comment (isovector) [reply src/Web/Suavemente/Input.hs]: Maybe we can't get our hands on it as written today, but `tr`s definitely support classes and the `visibility` property

**Comment by isovector** [comment]
My sig `Suave a -> (a -> Bool) -> Suave b -> Suave (a, Maybe b)` is intending to use the `a` as the selector and the `a -> Bool` as a visibility control for the `b`. If we're going to provide this, it'd be bad if the HS can branch on things hidden by the JS, I think?

**Comment by tonyday567** [comment]
Famous last words, but what can go wrong, since we have a lawful Applicative instance? If I have a Suave Bool and a Suave a, I can make a Suave (Maybe a).  maybeInput is then all about making the effect work as intended, where all type bets are off; but the haskell side is done and dusted!

It sounds like the `Suave b` you have in mind is the checkboxShow that's already inside the maybeInput?  Just looking at it as a tuple, we have `(,) <$> checkboxShow label cl start <*> div' cl (display start) sa` and wrapping it as inputMaybe just enforces the same visibility class name, and the same visibility at the start.

**Comment by isovector** [comment]
The concern is more at a conceptual level---you wouldn't expect hidden options on the JS side to affect rendering on the HS side, so let's just force that to be the case by lifting the `Maybe`.

**Comment by tonyday567** [comment]
I changed it a touch to remove the maybe aspect, which I think was confusing the issues.  On the effect side, the input still exists,, it's just that it's display happens to be hidden.

**Comment by isovector** [comment]
Hi! I'm travelling right now and haven't gotten much of a chance to wrap my brain around this yet. Apologies for the delay, but please don't think I've forgotten. Back in five days.

**Comment by isovector** [comment]
I'm happy with this as the library stands today. I think a better long-term solution would be as you say to add some more structure to the output. The `toggleInput` combinator could do some sort of SYB to inject classes onto its contained table rows, and avoid the `textBox_` div hack.

That's a change I'm not up to making today, but I'll file it under the "one day" projects. In the meantime this PR addresses the issue well, so I'm happy to merge it. Thanks for your patience!

**Comment by tonyday567** [comment]
Thanks and no worries about the delay.  This is open source!

**Pull request merged**